### DISCUSSION TITLE
将简悦 · 同步助手部署在远程（ SaasS 化 ），方便多个扩展端共用一个同步助手

### DISCUSSION THREAD

**Discussion by shadu120** [Kenshin/simpread#3898] [同步助手]
> 把简悦同步助手部署在远程，让你多个设备共享一个同步助手，即可实现自动同步。

## 为什么要这么干？

我有三台设备在使用：办公室一台 Mac，家里一台 Mac，出差路上一个 Windows，我需要在三个电脑上都安装客户端 + 坚果云才能实现自动同步。

所以，为什么不把 **同步助手  SaaS** 化了呢？

### 演示环境准备

![image](https://user-images.githubusercontent.com/3023471/168279137-20af5694-a488-4f68-8c9a-0d53ee18a19e.png)

#### 准备工作1：端口转发

先做个配置，确保简悦浏览器所在终端电脑本机 127.0.0.1 的 7026 端口映射到了同步助手所在的远程服务 192.168.200.226 的 7026 端口。
映射工具推荐：haproxy，可后台服务模式运行。

**为了方便，本次演示，我是用 SSH tunnel 做的端口转发。**

MacBook Air 127.0.0.1:7026 ---> 192.168.200.226:7026
Windows 11  127.0.0.1:7026 ---> 192.168.200.226:7026

虽然端口转发不属于本文档主要内容，但是为了方便大家快速体验，我做个小介绍。

-  Mac系统因为大都数本身已经自带ssh服务，使用：`ssh -fN -L 7026:192.168.22.226:7026 username@127.0.0.1` 就可以把本机的7026端口映射到远程服务器192.168.22.206的7026。

- Windows系统，网上也有netsh的操作方法，有失败概率。可以使用github上这款工具[lcx](https://github.com/rayschu/lcx)，命令也很简单：`python lcx.py -l 127.0.0.1 -p 7026 -r 192.168.200.226 -P 7026`。当然你搜索：端口转发、PortMapper等关键会有更多高级工具。

#### 准备工作2：在两台电脑上安装好简悦客户端

分别安装简悦Chrome客户端以及插件……

#### 第一步：在远程Windows中启动简悦.同步助手客户端，并设置好存储目录
本演示的Windows主机 IP 为：192.168.222.26，运行在虚拟机中。

![image](https://user-images.githubusercontent.com/3023471/168279429-79ea93e6-efaf-4a16-a8c7-ef62505a4fd8.png)

#### 第二步：配置Windows电脑的简悦

Windows A-ChromeA-SimleRead A，先完成配置（激活到高级版）
然后通过插件设置，连接同步助手，完成授权。

![image](https://user-images.githubusercontent.com/3023471/168274960-3b09ed87-74f9-48fb-b12a-8cf37a3b355a.png)

之后，开启本地助手的自动同步功能。动动手，选择通过简悦助手覆盖本地配置文件。

![image](https://user-images.githubusercontent.com/3023471/168276496-8082c6d7-9f99-4a24-bee4-abbf5bb51709.png)

#### 第三步：配置 MacBook Air 电脑的简悦

从 Windows Chrome 中导出简悦的配置文件，传到 Mac 电脑后，导入浏览器B中的简悦。

![image](https://user-images.githubusercontent.com/3023471/168275676-00b4a2c9-ddd9-43f6-864a-f126b1601347.png)

在 Chrome B，SimleRead B 中，授权简悦 · 同步助手，并启动自动同步功能。

![image](https://user-images.githubusercontent.com/3023471/168276056-205d13ed-eb14-4d22-9258-8adf058d99f6.png)

基本搞定了！

### 效果一：稍后读数据同步

在 Windows 电脑中，使用简悦查看网文，并保存到稍后阅读。
![image](https://user-images.githubusercontent.com/3023471/168277638-b8c57ec7-a67a-42fe-b64c-2bac8017f19a.png)

通过 Mac 电脑访问简悦的稍后阅读。
![2022-05-13_18-40-44](https://user-images.githubusercontent.com/3023471/168277789-04139b1d-bdf3-4b6f-acb9-cbeeac918336.png)

Mac 电脑中标注的文章，在 Windows 稍后阅读中同样可浏览。

![2022-05-13_18-36-18](https://user-images.githubusercontent.com/3023471/168278317-c5423ae5-8654-49d5-af6a-c3c00cc3dfbb.png)

![image](https://user-images.githubusercontent.com/3023471/168278360-eee96489-eba6-4a9a-9440-9f93f930e878.png)

哈哈，数据同步了，搞定！

### 效果二：离线存储文件同步
通过 Windows 或 Mac 电脑，阅读网文后，导出为离线文档，在远端的同步助手 output 文件夹，你能看到想要的文件。
![image](https://user-images.githubusercontent.com/3023471/168278206-aa571618-e548-4e9f-a230-6c0cdc385632.png)

## 最后

感谢简悦这款优秀的产品，我非常喜欢！

这个使用方法有一个问题是：你的浏览器所在终端必须通过端口映射才能使用远程的同步助手。希望简悦未来版本迭代中，允许客户端配置同步助手的IP地址，这样就不需要大费周折啦！

**Comment by Kenshin** [comment]
哈哈，很有意思的文章，我补充几点。

## 优点

1. 通过此方式，可以只安装扩展端，通过端口和 IP 的映射方案实现 `n:1` 的方式，节省了每台设备都要安装同步助手的步骤。
2. 如果你因为公司缘故无法安装同步助手，可以使用此教程来实现。

## 注意事项

1. 此方式因为需要使用高级账户权限，所以最多只限三台设备。

**Reply by michsiu** [reply]
想问下这个方法需要远端电脑保持一致开启吗

**Comment by shaoye1995** [comment]
要是可以更简单部署就好了比如装到docker里

**Reply by Kenshin** [reply]
哈哈，会考虑的。

**Reply by ddkk3000** [reply]
我也赞同，如果能直接布置在docker里这种方式也许更实用。
